File: scripts/debugiso.py
"""Debug: dump the top-level ISO directory to understand its layout."""
import struct

ISO = "C:/Users/meyer.n/Downloads/Sims2.iso"

with open(ISO, 'rb') as f:
    # Check all volume descriptors
    for sector in range(16, 32):
        f.seek(sector * 2048)
        vd = f.read(2048)
        vd_type = vd[0]
        magic = vd[1:6]
        if magic != b'CD001':
            break
        types = {0: 'boot', 1: 'primary', 2: 'supplementary', 255: 'terminator'}
        print(f"Sector {sector}: type={vd_type} ({types.get(vd_type,'?')})")
        if vd_type == 1:  # Primary Volume Descriptor
            root_lba  = struct.unpack_from('<I', vd, 158)[0]
            root_size = struct.unpack_from('<I', vd, 166)[0]
            print(f"  Root dir: lba={root_lba}, size={root_size}")
            # Dump top-level entries
            f.seek(root_lba * 2048)
            data = f.read(root_size)
            pos = 0
            count = 0
            while pos < len(data) and count < 50:
                rec_len = data[pos]
                if rec_len == 0:
                    pos = (pos // 2048 + 1) * 2048
                    if pos >= len(data): break
                    continue
                flags = data[pos + 25]
                file_lba  = struct.unpack_from('<I', data, pos + 2)[0]
                file_size = struct.unpack_from('<I', data, pos + 10)[0]
                name_len  = data[pos + 32]
                name = data[pos + 33:pos + 33 + name_len]
                is_dir = bool(flags & 2)
                print(f"  {'DIR' if is_dir else 'FILE':4s} {name!r:40s}  lba={file_lba}  size={file_size}")
                pos += rec_len
                count += 1

File: scripts/debugIso.ts
import { closeSync, openSync } from "node:fs";
import { readAt, SECTOR_SIZE, listDirectory } from "./lib/iso9660.ts";

// Dumps every volume descriptor on a disc image, and the top-level directory
// of its primary (non-Joliet) tree — the shape a reader sees before it knows
// to look for Joliet's long names. Useful for a first look at an unfamiliar
// disc's layout.
const DESCRIPTOR_NAMES: Record<number, string> = { 0: "boot", 1: "primary", 2: "supplementary", 255: "terminator" };

function main(): void {
    const isoPath = process.argv[2];
    if (!isoPath) {
        console.error("usage: debugIso.ts <path-to-iso>");
        process.exitCode = 1;
        return;
    }

    const fd = openSync(isoPath, "r");
    try {
        for (let sector = 16; sector < 32; sector += 1) {
            const descriptor = readAt(fd, sector * SECTOR_SIZE, SECTOR_SIZE);
            if (descriptor.length < 6 || descriptor.toString("ascii", 1, 6) !== "CD001") {
                break;
            }
            const type = descriptor[0]!;
            console.log(`Sector ${sector}: type=${type} (${DESCRIPTOR_NAMES[type] ?? "?"})`);

            if (type === 1) {
                const rootLba = descriptor.readUInt32LE(158);
                const rootSize = descriptor.readUInt32LE(166);
                console.log(`  Root dir: lba=${rootLba}, size=${rootSize}`);

                const entries = listDirectory(fd, rootLba, rootSize).slice(0, 50);
                for (const entry of entries) {
                    console.log(
                        `  ${entry.isDirectory ? "DIR " : "FILE"} ${entry.name.padEnd(40)}  lba=${entry.lba}  size=${entry.size}`,
                    );
                }
            }
        }
    } finally {
        closeSync(fd);
    }
}

main();
